### PR TITLE
[test-operator] Fix handling of failed job

### PR DESCRIPTION
**Pull request by lpiwowar**
[test-operator] Fix handling of failed job

There are two issues in the way of how the role does handle failed job

- Currently, the role waits for the Job that starts the execution of
  the test pod to reach Completed state but this state is only reached
  when the execution of the tests was successful. If there is an error
  the job finishes with Failed state.

- Logs are not collected when the the test pod times out

This patch updates the role so that it correctly waits for the job
completion (either with success or failure) and ensures that we collect
the logs (output of oc logs) even when the test pod times out.

As a pull request owner and reviewers, I checked that:
- [x] Appropriate testing is done and actually running
- [x] Appropriate documentation exists and/or is up-to-date:
  - [x] README in the role
  - [x] Content of the docs/source is reflecting the changes

### CODE REVIEW

**Review by kopecmartin — APPROVED** [review]
lgtm

File: roles/test_operator/tasks/main.yml
Comment (marios) [inline roles/test_operator/tasks/main.yml:146]: ok default 3600
 https://github.com/openstack-k8s-operators/ci-framework/blob/fa2fcee6fec40f7ede3f2868a368149c583c51d5/roles/test_operator/defaults/main.yml#L23

File: roles/test_operator/tasks/main.yml
Comment (marios) [inline roles/test_operator/tasks/main.yml:157]: is there nothing returned for us here from the tempest.resources ? feels a bit fragile the way you are determining the timeout

Comment (lpiwowar) [reply roles/test_operator/tasks/main.yml:157]: As we've already discussed there is unfortunately probably no other way how to determine the timeout.

File: roles/test_operator/tasks/main.yml
Comment (queria) [inline roles/test_operator/tasks/main.yml:163]: would this mean there will be no logs collected in case it hanged/timed-out?

just wild guessing but if its maybe due to access to persistent storage or such, could there be some read-only parallel access, or maybe force terminate the execution on timeout to get access to logs?

Comment (lpiwowar) [reply roles/test_operator/tasks/main.yml:163]: I wanted to skip these because there are no longs to be collected in case of time out [1]. The logs are moved to the directory where the PV is mounted at the end of the test execution.

However, it is possible that there might be an issue with parallel access to the logs. 

[1] https://github.com/openstack-k8s-operators/tcib/blob/106264ab0a0b18be278d69b0664dad6a4f9be29a/container-images/tcib/base/os/tempest/run_tempest.sh#L265

**Review by marios — APPROVED** [review]
as discussed yesterday this won't affect any currently running job until we wire it up so lets merge to unblock work here

**Comment by SeanMooney** [comment]
i used depend on to include it in https://github.com/openstack-k8s-operators/nova-operator/pull/665
and it seamed to function fine. that last run did not time out but the logs are present
https://logserver.rdoproject.org/65/665/8a814593c6acbbf86b6ad3022bb565bb23d73247/github-check/nova-operator-tempest-multinode/ac65ef4/controller/ci-framework-data/tests/test_operator/stestr_results.html
so this is at least not breaking that existing behaivor

File: roles/test_operator/tasks/main.yml
Comment (SeanMooney) [reply roles/test_operator/tasks/main.yml:163]: that is not correct

generate logs is post processing the tempest logs.

the tempest logs are stored in a binary format by stester when tempest is run in parrall with the execution fo the these

this https://github.com/openstack-k8s-operators/tcib/blob/106264ab0a0b18be278d69b0664dad6a4f9be29a/container-images/tcib/base/os/tempest/run_tempest.sh#L226C5-L226C66

stestr last --subunit > ${TEMPEST_PATH}testrepository.subunit

is just copying the last log to a new location.

however what we should do doing is making sure that the default .stestr folder
is mind mounted to a log volume so that the logs are directly written to a PVC and we can post process them even when a time out happens.


alternatively we shoudl be passing the time out to the tempest run invocation

timeout ${TEMPEST_TIMEOUT} tempest run ${TEMPEST_ARGS}

hhttps://github.com/openstack-k8s-operators/tcib/blob/106264ab0a0b18be278d69b0664dad6a4f9be29a/container-images/tcib/base/os/tempest/run_tempest.sh#L215


then the job will finish in a roghyly fixed time and it will generate the logs properly then we done need to wait for a fix period of time here in this script.

Comment (SeanMooney) [reply roles/test_operator/tasks/main.yml:163]: making the run_tempest.sh script enforce the timeout when invoking tempest is obviously the simpler option

**Pull request merged**
